tfimm/architectures/lora/README.md

```diff
@@ -0,0 +1,175 @@
+by passing the values `"none"`, `"lora_only"` or  `"all"` for the `lora_train_bias` 
+parameter.
+
+```python
+lora_model = convert_to_lora_model(
+    model, lora_rank=2, lora_train_bias="lora_only"
+)
+```
+
+## Sequential and functional models
+
+The current implementations focusses on models created by subclassing, which is the case
+for all `tfimm` models. In particular, the registry system works only for subclassed
+models. However, some of the functionality also works for functional models.
+
+- The LoRA layers are the basic building blocks for both subclassed as well as 
+  functional models.
+- Transferring weights works for all models, regardless of type, provided the regular
+  model and LoRA variant have the same architecture with the exception of LoRA layers.
+  Use the function to tranfer weights to LoRA.
+  ```python
+  from tfimm.architectures import lora
+  from tfimm.models import transfer_weights
+  
+  # Transfer weights into the LoRA model
+  transfer_weights(
+      regular_model, lora_model, weights_to_ignore=lora.LORA_WEIGHT_NAMES)
+  ```
+- After training, we need to manually merge weights and then transfer them back to the
+  regular model.
+  ```python
+  lora.merge_weights(lora_model)
+  transfer_weights(lora_model, regular_model)
+  ```
+- The functions `lora_trainable_weights` and `lora_non_trainable_weights` work for all
+  models, regardless of type and return a list of weights to be used for LoRA training
+  (or all other weights).
+
+## Comparison to PyTorch implementation
+
+The official PyTorch implementation can be found 
+[here](https://github.com/microsoft/LoRA/tree/main). The differences between TensorFlow
+and PyTorch mean that we cannot simply replicate the PyTorch implementation.
+
+- In PyTorch it is possible to mark only LoRA parameters as trainable via
+  ```python
+  # This sets requires_grad to False for all parameters without the 
+  # string "lora_" in their names
+  lora.mark_only_lora_as_trainable(model)
+  ```
+  In TensorFlow it is _not_ possible to change the `trainable` attribute of a variable
+  after the variable has been created. Instead, we provide the function 
+  `lora_trainable_weights`, which provides a list of weights to be used for LoRA 
+  training. For `tfimm` architectures we also override the model's `trainable_weights`
+  property, so the model can be used with Keras' built-in `model.fit()`.
+- PyTorch allows easy saving and loading a partial state of the model via `state_dict`,
+  ```python
+  torch.save(lora.lora_state_dict(model), "ckpt_lora.pt")
+  model.load_state_dict(torch.load("ckpt_lora.pt"), strict=False)
+  ```
+  In TensorFlow `model.save_weights()` does not allow saving or loading partial states.
+  Something similar can most likely be achieved using TensorFlow checkpoints, but we
+  haven't investigated it yet. In particular, currently we first have to load the
+  regular model into memory and then transfer its weights to the LoRA version.
+
```

Comment:
could we add a list of currently implemented LoRA layers and LoRA models?

Reply:
Yes, added that.